### DISCUSSION TITLE
Cannot mutate userdata from a field on "root" userdata

### DISCUSSION THREAD

**Discussion by SeanOMik** [mlua-rs/mlua#466] [Q&A]
I have a struct passed to Lua as userdata. It has a field that is also returned to lua as userdata. That field has a method that mutates it's self. When the method on it is used, it is not modify the field in the "root" userdata like I expected. Here's some code that would show what I mean:

```lua
-- c variable is provided by Rust host to Lua.
-- type is `Camera` userdata.

-- c.transform is `Transform` userdata
c.transform:translate(0, 0.15, 0)
print("Moving camera to: " .. tostring(c.transform))
```
The above code prints the original value, the transform is not modified. In the below code, the transform is modified:
```lua
local t = c.transform 
t:translate(0, 0.15 * dt, 0)
c.transform = t
```

I believe that its because I'm using `add_field_method_get` in the `UserData` trait impl of `Camera` to add a field to get `transform`. In the method, I copy `Transform` before returning it to Lua. How could I return a reference?

**Comment by CrashAndSideburns** [comment]
I'm trying to embed Lua into an emulator that I'm working on, and I'm running into the exact same issue. Is there any chance that you've resolved this? I can't seem to find any documentation discussing how this should be done, or find anything similar being done in examples.

**Reply by khvzak** [reply]
There are couple of options:
1) Wrap nested struct to `Rc<RefCell<_>>` then it can be mutated when cloned
2) Keep nested struct in AnyUserData from the beginning (cloning it only creates a new references to existing object)

**Reply by CrashAndSideburns** [reply]
Thanks! I just got it working by wrapping the inner struct in `Rc<RefCell<_>>`. I don't suppose that there would be a way to do this without wrapping every single field of the struct in this? I would need to make some pretty big changes to the code base of my emulator to do that. It feels like it should be possible to just wrap the outer struct, but maybe I'm wrong? For instance, I can do
```rust
struct Foo {
    inner: BarWrapper,
}

struct BarWrapper(Rc<RefCell<Bar>>);

struct Bar {
    value: String
}
```
but would it be possible to wrap `Foo` instead of `Bar`? Sorry if this isn't explained well, I can try to put together a sample from my code if this isn't helpful.

**Reply by khvzak** [reply]
You can wrap only `Foo`, but then you need to create few other structs, like `BarWrapper(Rc<RefCell<Foo>>)` to attach different methods to them.
Eg. when you call `bar:update()` the code should know that `inner` field of `Foo` needs to be updated.

**Reply by CrashAndSideburns** [reply]
Ah, thanks! That works. That's going to end up being a lot simpler than overhauling all of the types in the core of my emulator, and it should help decouple the Lua interfaces from the actual code. Thanks so much!
